src/Neo/Network/P2P/Payloads/Signer.cs

```diff
@@ -66,6 +66,33 @@ public class Signer : IInteroperable, ISerializable
             /*AllowedGroups*/       (Scopes.HasFlag(WitnessScope.CustomGroups) ? AllowedGroups.GetVarSize() : 0) +
             /*Rules*/               (Scopes.HasFlag(WitnessScope.WitnessRules) ? Rules.GetVarSize() : 0);
 
+        public bool Equals(Signer other)
+        {
+            if (ReferenceEquals(this, other))
+                return true;
+            return Account == other.Account &&
+                Scopes == other.Scopes &&
+                AllowedContracts.SequenceEqual(other.AllowedContracts) &&
+                AllowedGroups.SequenceEqual(other.AllowedGroups) &&
+                Rules.SequenceEqual(other.Rules);
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null)
+                return false;
+
+            if (obj is not Signer signerObj)
+                return false;
+            else
+                return Equals(signerObj);
```

Comment:
`return obj as Signer signer && Equals(signer)` is more simple.

Reply:
Need to check for `null` too, thats why.

Reply:
> Need to check for `null` too, if it can't convert to `signer`, thats why.

`as` is null safe

Reply:
Agree with @nan01ab

Reply:
> `as` is null safe

It is not null safe, Im not changing it.

Reply:
> > `as` is null safe
> 
> It is not null safe, If it can't convert to the type it will be `null`.
> 
> The `as` operator considers only reference, nullable, boxing, and unboxing conversions. You can't use the `as` operator to perform a user-defined conversion. To do that, use a [cast expression](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#cast-expression).
> 
> Reference https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#as-operator

But `Signer` is a `class`, and obj is an `object`. why isn't this null safe?

Reply:
no, if it can't convert to type it will be `null`.

Reply:
> no, if `object` can't convert to type it will be `null`.

`is`, not `as`, there is a typo

Reply:
It is pretty safe to use `is` here, can you please provide a canse where object can not convert into a type?

Following test is true. 
```C#
           object obj = null;
            var isSigner = obj is Signer signer; // false, obj is null

            obj = new Signer();
            var isSigner2 = obj is Signer signer2; // true, obj is Signer

            obj = new UInt160();
            var isUInt160 = obj is UInt160 uint160; // true, obj is UInt160

            obj = "string";
            var isString = obj is string str; // true, obj is string

            obj = 123;
            var isInt = obj is int number; // true, obj is int

            obj = new ECPoint();
            var isECPoint = obj is ECPoint ecPoint; // true, obj is ECPoint

            obj = new WitnessRule();
            var isWitnessRule = obj is WitnessRule witnessRule; // true, obj is WitnessRule

            obj = new object();
            var isNotConvertible = obj is Signer notSigner; // false, obj is not Signer

            Assert.IsTrue(!isSigner && isSigner2 && isUInt160 && isString && isInt && isECPoint && isWitnessRule && !isNotConvertible);
```